### PR TITLE
Feature ETP-713: Added base builder for the toolbar

### PR DESCRIPTION
**Pull request by Sankku**

### CODE REVIEW

File: src/com/etendoerp/metadata/MetadataServlet.java [hunk 1]
```diff
@@ -42,9 +43,84 @@ private void handleRequest(HttpServletRequest request, HttpServletResponse respo
...
+                sendErrorResponse(response, HttpServletResponse.SC_BAD_REQUEST, "Invalid toolbar path");
+                return;
+            }
+            try {
+                String windowId = pathParts[2];
+                String tabId = (pathParts.length >= 4 && !"undefined".equals(pathParts[3])) ? pathParts[3] : null;
+                JSONObject toolbar = fetchToolbar(windowId, tabId);
+                sendSuccessResponse(response, toolbar);
+            } catch (IllegalArgumentException e) {
+                sendErrorResponse(response, HttpServletResponse.SC_NOT_FOUND, e.getMessage());
```
Comment (LuuchoRocha) [inline src/com/etendoerp/metadata/MetadataServlet.java:58]: Please, create a new exception called `UnprocessableContentException`.
Remove this sendErrorResponse and instead use:

`throw new UnprocessableContentException(e.getMessage());`

you will need to update the `getResponseStatus` of `BaseServlet`:


```
    protected int getResponseStatus(Exception e) {
        Object clazz = e.getClass();

        if (clazz.equals(OBSecurityException.class) || clazz.equals(UnauthorizedException.class)) return 401;
        else if (clazz.equals(MethodNotAllowedException.class)) return 405;
        else if (clazz.equals(UnprocessableContentException.class)) return 422;
        else return 500;
    }
```

File: src/com/etendoerp/metadata/MetadataServlet.java [hunk 2]
```diff
@@ -42,9 +43,84 @@ private void handleRequest(HttpServletRequest request, HttpServletResponse respo
...
+            } catch (Exception e) {
+                sendErrorResponse(response, HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Error processing request");
```
Comment (LuuchoRocha) [inline src/com/etendoerp/metadata/MetadataServlet.java:60]:
```suggestion
                new InternalServerException();
```

File: src/com/etendoerp/metadata/builders/ToolbarBuilder.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        this.connectionProvider = ConnectionProviderContextListener.getPool();
+    }
+
+    public JSONObject toJSON() {
+        try {
+            Window window = OBDal.getInstance().get(Window.class, windowId);
+            return buildToolbarJSON(window);
+        } catch (Exception e) {
+            logger.error("Error building toolbar for window: {}", windowId, e);
+            throw new RuntimeException("Failed to build toolbar", e);
```
Comment (LuuchoRocha) [inline src/com/etendoerp/metadata/builders/ToolbarBuilder.java:37]: Please use more specific exceptions

File: src/com/etendoerp/metadata/builders/ToolbarBuilder.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    private JSONObject createButtonJSON(ButtonConfig config) throws Exception {
+        JSONObject button = new JSONObject();
+        button.put("id", config.id);
+        button.put("name", Utility.messageBD(connectionProvider, config.name, language));
+        button.put("action", config.action);
+        button.put("enabled", config.enabled);
+        button.put("visible", true);
+        button.put("icon", config.icon);
+        return button;
+    }
```
Comment (LuuchoRocha) [inline src/com/etendoerp/metadata/builders/ToolbarBuilder.java:66]:
```suggestion
    private JSONObject createButtonJSON(ButtonConfig config) throws Exception {
        JSONObject button = new JSONObject();

        button.put("id", config.id);
        button.put("name", Utility.messageBD(connectionProvider, config.name, language));
        button.put("action", config.action);
        button.put("enabled", config.enabled);
        button.put("visible", true);
        button.put("icon", config.icon);

        return button;
    }
```

File: src/com/etendoerp/metadata/builders/ToolbarBuilder.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        buttons.put("GRID_VIEW", new ButtonConfig(
+                "GRID_VIEW",
+                "OBUIAPP_GridView",
+                "GRID_VIEW",
+                true,
+                "grid"
+        ));
+
+        return buttons;
+    }
```
Comment (LuuchoRocha) [inline src/com/etendoerp/metadata/builders/ToolbarBuilder.java:154]:
```suggestion
    private Map<String, ButtonConfig> getStandardButtons() {
        Map<String, ButtonConfig> buttons = new HashMap<>();

        buttons.put("NEW", new ButtonConfig(
                "NEW",
                "OBUIAPP_NewDoc",
                "NEW",
                true,
                "plus"
        ));
        buttons.put("SAVE", new ButtonConfig(
                "SAVE",
                "OBUIAPP_SaveRow",
                "SAVE",
                true,
                "save"
        ));
        buttons.put("DELETE", new ButtonConfig(
                "DELETE",
                "OBUIAPP_DeleteRow",
                "DELETE",
                !isNew,
                "trash"
        ));
        buttons.put("REFRESH", new ButtonConfig(
                "REFRESH",
                "OBUIAPP_RefreshData",
                "REFRESH",
                true,
                "refresh-cw"
        ));
        buttons.put("FIND", new ButtonConfig(
                "FIND",
                "OBUIAPP_Find",
                "FIND",
                true,
                "search"
        ));
        buttons.put("EXPORT", new ButtonConfig(
                "EXPORT",
                "OBUIAPP_ExportGrid",
                "EXPORT",
                true,
                "download"
        ));
        buttons.put("ATTACHMENTS", new ButtonConfig(
                "ATTACHMENTS",
                "OBUIAPP_Attachments",
                "ATTACHMENTS",
                true,
                "paperclip"
        ));
        buttons.put("GRID_VIEW", new ButtonConfig(
                "GRID_VIEW",
                "OBUIAPP_GridView",
                "GRID_VIEW",
                true,
                "grid"
        ));

        return buttons;
    }
```

**Review by LuuchoRocha — CHANGES_REQUESTED** [review]

File: src/com/etendoerp/metadata/MetadataServlet.java [hunk 6]
```diff
@@ -46,23 +48,21 @@ private void handleRequest(HttpServletRequest request, HttpServletResponse respo
         } else if (path.startsWith("/toolbar")) {
             String[] pathParts = path.split("/");
             if (pathParts.length < 3) {
-                sendErrorResponse(response, HttpServletResponse.SC_BAD_REQUEST, "Invalid toolbar path");
-                return;
+                throw new UnprocessableContentException("Invalid toolbar path");
```
Comment (LuuchoRocha) [inline src/com/etendoerp/metadata/MetadataServlet.java:51]: actually this should be a not found 404 :nerd_face:

**Review by LuuchoRocha — APPROVED** [review]

**Review by LuuchoRocha — APPROVED** [review]

**Review by Agustin-Calderon — APPROVED** [review]

**Pull request merged**